packages/execute-driver/lib/plugin.js

```diff
@@ -0,0 +1,134 @@
+import BasePlugin from '@appium/base-plugin';
+import _ from 'lodash';
+import path from 'path';
+import cp from 'child_process';
+import B from 'bluebird';
+
+const FEAT_FLAG = 'execute_driver_script';
+const DEFAULT_SCRIPT_TIMEOUT = 1000 * 60 * 60; // default to 1 hour timeout
+const SCRIPT_TYPE_WDIO = 'webdriverio';
+
+export default class ExecuteDriverPlugin extends BasePlugin {
+
+  static newMethodMap = {
+    '/session/:sessionId/appium/execute_driver': {
+      POST: {
+        command: 'executeDriverScript',
+        payloadParams: {required: ['script'], optional: ['type', 'timeout']}
+      }
+    }
+  };
+
+  /**
+   * This method takes a string which is executed as javascript in the context of
+   * a new nodejs VM, and which has available a webdriverio driver object, having
+   * already been attached to the currently running session.
+   *
+   * @param {string} script - the string representing the driver script to run
+   * @param {string} [scriptType='webdriverio'] - the name of the driver script
+   * library (currently only webdriverio is supported)
+   *
+   * @returns {Object} - a JSONifiable object representing the return value of
+   * the script
+   */
+  async executeDriverScript (next, driver, script, scriptType = 'webdriverio',
+    timeout = DEFAULT_SCRIPT_TIMEOUT) {
+
+    if (!driver.isFeatureEnabled(FEAT_FLAG)) {
+      throw new Error(`Execute driver script functionality is not available ` +
+                      `unless server is started with --allow-insecure including ` +
+                      `the '${FEAT_FLAG}' flag, e.g., --allow-insecure=${FEAT_FLAG}`);
+    }
+
+    if (scriptType !== SCRIPT_TYPE_WDIO) {
+      throw new Error(`Only the '${SCRIPT_TYPE_WDIO}' script type is currently supported`);
+    }
+
+    if (!driver.serverHost || !driver.serverPort) {
+      throw new Error('Address or port of running server were not defined; this ' +
+                      'is required. This is probably a programming error in the driver');
+    }
+
+    if (!_.isNumber(timeout)) {
+      throw new Error('Timeout parameter must be a number');
+    }
+
+    const driverOpts = {
+      sessionId: driver.sessionId,
+      protocol: 'http', // Appium won't ever be behind ssl locally
+      hostname: driver.serverHost,
+      port: driver.serverPort,
+      path: driver.serverPath,
+      isW3C: true,
+      isMobile: true,
+      capabilities: driver.caps
+    };
+    this.logger.info(`Constructed webdriverio driver options; W3C mode is ${driverOpts.isW3C ? 'on' : 'off'}`);
```

Comment:
I see isW3C is always set to true above, how it could be turned off then?

Reply:
in this context, `isW3C` just ensures that the w3c commands are available on the driver. because we know appium is w3c compatible, this shouldn't backfire?